data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+
+In many imperative languages, the `spicy_cat ("hello", "world")` syntax reads as a function call with two parameters; but in OCaml, it denotes applying the function `spicy_cat` to a pair containing `"hello"` and `"world"`.
+
+### Currying and Uncurrying
+
+In the previous sections, two kinds of multiple-parameter functions have been presented.
+- Functions returning a function, such as `sweet_cat` and `sour_cat`
+- Functions taking a tuple as a parameter, such as `spicy_cat`
+
+Interestingly, both kinds of functions provide a way to pass several pieces of data to a body while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. It is always possible to translate back and forth between functions that look like `sweet_cat` (or `sour_cat`) and functions that look like `spicy_cat`.
+
+These translations have names:
+* [Currying](https://en.wikipedia.org/wiki/Currying) goes from the `spicy_cat` form into the `sour_cat` (or `sweet_cat`) form.
+* Uncurrying goes from the `sour_cat` (or `sweet_cat`) form into the `spicy_cat` form.
+
+It also said that `sweet_cat` and `sour_cat` are _curried_ functions whilst `spicy_cat` is _uncurried_.
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry).
+
+From a typing perspective, this means the following types are equivalent:
+- `string -> (string -> string)` &mdash; curried function type
+- `string * string -> string` &mdash; uncurried function type
+
+Here, this is shown using `string` as an example, but it applies to any group of three types.
+
+We will not dive any deeper into the details here, but these equivalences can be formally defined using _ad-hoc_ mathematics.
+
+It is rarely necessary to use the functions above. However, it is important to understand that changing a function may not need any code refactoring. A function can be given several equivalent forms and changed into another, either using refactoring or using a higher-order function. Since functions are values, currying and uncurrying are operations on those values.
+
+In practice, curried functions are the default form functions should take because:
+- They allow partial application
+- Less editing, no parentheses or commas
+- No pattern matching over a tuple
+
+## Functions With Side-Effects
+
+With respect to its type, a function is processes input data from its *domain*, the set of input values, and produce a result data from its *codomain*, the set of output values.
+```ocaml
+# string_of_int;;
+- : int -> string = <fun>
+```
+The function `string_of_int`
+* Has domain `int`
+* Has codomain `string`
```

Comment:
So is domain / codomain like types? Because `int` and `string` are types, but they're also domains and codomains?